routing/pathfind.go

```diff
@@ -585,10 +590,79 @@ func findPaths(tx *bolt.Tx, graph *channeldb.ChannelGraph,
 		candidatePaths pathHeap
 	)
 
+	var err error
+	if tx == nil {
+		tx, err = graph.Database().Begin(false)
+		if err != nil {
+			return nil, err
+		}
+		defer tx.Rollback()
+	}
+
+	// Iterate over all nodes in the graph database, and add them to an
+	// array which can back an in-memory graphSource. We do this to avoid
+	// having to read from the database during each call, speeding up the
+	// execution significantly.
+	var nodes []*channeldb.LightningNode
+	if err := graph.ForEachNode(nil,
+		func(_ *bolt.Tx, node *channeldb.LightningNode) error {
+			nodes = append(nodes, node)
+			return nil
+		}); err != nil {
+		return nil, err
+	}
+
+	// To the same for the edges, mapping each vertex to a list of its
+	// edges.
+	edges := make(map[vertex][]edge)
+	for _, node := range nodes {
+		err := node.ForEachChannel(tx, func(tx *bolt.Tx,
+			edgeInfo *channeldb.ChannelEdgeInfo,
+			outEdge, inEdge *channeldb.ChannelEdgePolicy) error {
+			v := newVertex(node.PubKey)
+			e := edge{
+				edgeInfo: edgeInfo,
+				outEdge:  outEdge,
+				inEdge:   inEdge,
+			}
+			edges[v] = append(edges[v], e)
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Now let nodes and edges back an in-memory graphSource, for quickly
+	// iterating through this graph for the repeated calls to its methods.
+	g := &graphSource{
```

Comment:
All of this code to construct the `graphSource` looks like it can be pulled out cleanly into its own function `makeMemoryGraph` or something. This would fit well with the idea of making `graphSource` an interface instead of a struct too.